File: app/src/main/java/me/devsaki/hentoid/viewholders/DrawerItem.java
package me.devsaki.hentoid.viewholders;

import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.items.AbstractItem;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.database.domains.Content;
import me.devsaki.hentoid.enums.AlertStatus;
import me.devsaki.hentoid.enums.Site;

import static androidx.core.view.ViewCompat.requireViewById;

public class DrawerItem extends AbstractItem<DrawerItem.DrawerViewHolder> {

    // Label of the item
    private final String label;
    // Icon of the item
    private final int icon;
    // Flag to indicate new content
    private boolean flagNew = false;
    // Flag to indicate an alert
    private AlertStatus alertStatus = AlertStatus.NONE;

    // Activity class to launch when clicking on the item
    private final Class<? extends AppCompatActivity> activityClass;
    // Corresponding site, if any
    private Site site;

    public DrawerItem(String label, int icon, Class<? extends AppCompatActivity> activityClass) {
        this.label = label;
        this.icon = icon;
        this.activityClass = activityClass;
    }

    public DrawerItem(Site site) {
        this.label = site.getDescription().toUpperCase();
        this.icon = site.getIco();
        this.activityClass = Content.getWebActivityClass(site);
        this.site = site;
    }

    public void setFlagNew(boolean flagNew) {
        this.flagNew = flagNew;
    }

    public void setAlertStatus(AlertStatus status) {
        this.alertStatus = status;
    }

    public void setSite(Site site) {
        this.site = site;
    }

    @Nullable
    public Site getSite() {
        return site;
    }

    public Class<? extends AppCompatActivity> getActivityClass() {
        return activityClass;
    }


    @NotNull
    @Override
    public DrawerItem.DrawerViewHolder getViewHolder(@NotNull View view) {
        return new DrawerViewHolder(view);
    }

    @Override
    public int getLayoutRes() {
        return R.layout.item_drawer;
    }

    @Override
    public int getType() {
        return R.id.drawer;
    }


    static class DrawerViewHolder extends FastAdapter.ViewHolder<DrawerItem> {

        private final ImageView icon;
        private final ImageView alert;
        private final TextView title;

        DrawerViewHolder(View view) {
            super(view);
            icon = requireViewById(view, R.id.drawer_item_icon);
            alert = requireViewById(view, R.id.drawer_item_alert);
            title = requireViewById(view, R.id.drawer_item_txt);
        }


        @Override
        public void bindView(@NotNull DrawerItem item, @NotNull List<Object> list) {
            icon.setImageResource(item.icon);
            if (item.alertStatus != AlertStatus.NONE) {
                alert.setVisibility(View.VISIBLE);
                alert.setColorFilter(ContextCompat.getColor(alert.getContext(), item.alertStatus.getColor()), android.graphics.PorterDuff.Mode.SRC_IN);
            } else alert.setVisibility(View.GONE);
            title.setText(String.format("%s%s", item.label, item.flagNew ? " *" : ""));
        }

        @Override
        public void unbindView(@NotNull DrawerItem item) {

        }
    }
}

File: app/src/main/java/me/devsaki/hentoid/viewholders/DrawerItem.kt
package me.devsaki.hentoid.viewholders

import android.graphics.PorterDuff
import android.graphics.Typeface
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.items.AbstractItem
import me.devsaki.hentoid.R
import me.devsaki.hentoid.enums.AlertStatus
import me.devsaki.hentoid.enums.Site

class DrawerItem<T>(
    val label: String,
    val icon: Int,
    uniqueId: Long,
    val italicFont: Boolean = false,
    private val mTag: T? = null
) :
    AbstractItem<DrawerItem.ViewHolder<T>>() {

    var site: Site? = null
    var flagNew = false
    var alertStatus = AlertStatus.NONE

    init {
        identifier = uniqueId
    }

    fun getObject(): T? {
        return mTag
    }

    override val type: Int get() = R.id.drawer

    override val layoutRes: Int get() = R.layout.item_drawer

    override fun getViewHolder(v: View): ViewHolder<T> {
        return ViewHolder(v)
    }

    class ViewHolder<T>(view: View) : FastAdapter.ViewHolder<DrawerItem<T>>(view) {
        private val icon: ImageView = itemView.findViewById(R.id.drawer_item_icon)
        private val alert: ImageView = itemView.findViewById(R.id.drawer_item_alert)
        private val title: TextView = itemView.findViewById(R.id.drawer_item_txt)

        override fun bindView(item: DrawerItem<T>, payloads: List<Any>) {
            icon.setImageResource(item.icon)
            if (item.alertStatus != AlertStatus.NONE) {
                alert.visibility = View.VISIBLE
                alert.setColorFilter(
                    ContextCompat.getColor(alert.context, item.alertStatus.color),
                    PorterDuff.Mode.SRC_IN
                )
            } else alert.visibility = View.GONE
            title.text = String.format("%s%s", item.label, if (item.flagNew) " *" else "")

            if (item.italicFont) title.setTypeface(null, Typeface.ITALIC)
            else title.setTypeface(null, Typeface.NORMAL)
        }

        override fun unbindView(item: DrawerItem<T>) {
            // Nothing special here
        }
    }

    companion object {
        fun fromSite(site: Site): DrawerItem<Site> {
            val result = DrawerItem(
                site.description.uppercase(),
                site.ico,
                site.code.toLong(),
                mTag = site
            )
            result.site = site
            return result
        }
    }
}